app/src/main/java/com/islaharper/dawnofandroid/presentation/screens/login/LoginScreen.kt

```diff
@@ -23,69 +29,118 @@ import androidx.compose.ui.tooling.preview.Preview
 import androidx.compose.ui.unit.dp
 import androidx.hilt.navigation.compose.hiltViewModel
 import androidx.navigation.NavHostController
+import com.google.android.gms.auth.api.identity.BeginSignInResult
+import com.google.android.gms.common.api.ApiException
 import com.islaharper.dawnofandroid.R
 import com.islaharper.dawnofandroid.domain.model.MessageBarState
+import com.islaharper.dawnofandroid.navigation.Screen
 import com.islaharper.dawnofandroid.presentation.common.GoogleButton
 import com.islaharper.dawnofandroid.presentation.components.MessageBar
+import com.islaharper.dawnofandroid.ui.theme.DawnOfAndroidTheme
+import com.islaharper.dawnofandroid.util.Resource
 
 @Composable
 fun LoginScreen(
     navHostController: NavHostController,
-    loginViewModel: LoginViewModel = hiltViewModel()
+    modifier: Modifier = Modifier,
+    loginViewModel: LoginViewModel = hiltViewModel(),
 ) {
-    val signedInState by loginViewModel.signedInState.collectAsState()
+    val launchOneTapState by loginViewModel.launchOneTapSignIn.collectAsState()
     val messageBarState by loginViewModel.messageBarState.collectAsState()
-    val apiResponse by loginViewModel.apiResponse.collectAsState()
+    val navigationState by loginViewModel.navigationState.collectAsState()
+    val oneTapSignInResponse by loginViewModel.oneTapSignInResponse.collectAsState()
 
     Scaffold(
         backgroundColor = MaterialTheme.colorScheme.surface,
         content = { contentPadding ->
             LoginContent(
-                modifier = Modifier.padding(contentPadding),
-                signedInState = signedInState,
+                modifier = modifier.padding(contentPadding),
+                signedInState = launchOneTapState,
                 messageBarState = messageBarState,
                 onButtonClicked = {
-                    loginViewModel.saveSignedInState(signedIn = true)
-                }
+                    loginViewModel.onGoogleButtonClick()
+                },
             )
-        }
+        },
     )
+
+    val authLauncher = authLauncher(loginViewModel)
+
+    if (launchOneTapState && oneTapSignInResponse is Resource.Success) {
```

Comment:
I think we should aim for a solution where we don't have to use state from the ViewModel to work out what logic to perform.

We seem to be using `launchOneTapState` for two purposes, one to indicate loading and one to indicate whether we should perform some one off event (like launching).

I think this can be simplified by making an extra `StateFlow` in the ViewModel for loading.

Something like :

```kotlin
val _loading: MutableStateFlow<Boolean>(false)
val loading: StateFlow<Boolean> = _loading
```

You will then have to set this to true or false within the ViewModel manually, or derive its value from some other StateFlow if this is easy to do.

The next step would be to expose a StateFlow for the `BeginSignInResult` which is what we actually need to launch the activity:

```kotlin
    private val _launchSignIn = MutableStateFlow<BeginSignInResult?>(null)
    val launchSignIn: StateFlow<BeginSignInResult?> = _launchSignIn
```

Now the logic in the Composable would become something like:

```kotlin
if (beginSignInResult != null) {
    intent = IntentSenderRequestBuilder(beginSignInResult.pendingIntent.intentSender.build()
   authLauncher.launch(intent)
   // set beginSignInResult back to null in the ViewModel to prevent one-off event from happening twice
}
```